server/src/routes/users.routes.ts

```diff
@@ -1,20 +1,118 @@
 import { Router } from 'express';
-import { sign } from 'jsonwebtoken';
+import { sign, verify } from 'jsonwebtoken';
+import { getRepository } from 'typeorm';
 import CreateUserService from '../services/CreateUserService';
 import UpdateUserService from '../services/UpdateUserService';
+import ResetUserPasswordService from '../services/ResetUserPasswordService';
 import DeleteUserService from '../services/DeleteUserService';
 import ensureAuthenticated from '../middlewares/ensureAuthenticated';
 import Queue from '../services/Queue';
 import SendConfirmAccountMail from '../jobs/SendConfirmAccountMail';
+import SendResetPasswordMail from '../jobs/SendResetPasswordMail';
 import authConfig from '../config/auth';
+import User from '../models/User';
 
 const usersRouter = Router();
 
+interface TokenPayload {
+  iat: number;
+  exp: number;
+  // FIXME: not sure why I can't call this 'user'
+  user: {
+    email: string;
+    id: string;
+    created_at: string;
+    updated_at: string;
+  };
+}
+
 /**
  * Repositories - business rules
  * Services - database
  */
 
+/**
+ * Endpoint for updating a user's password given a valid JWT token (user is obtained from the token payload)
+ */
+usersRouter.put('/reset-password/:token', async (request, response, next) => {
+  try {
+    const { token } = request.params;
+
+    const { newPassword, newPasswordConfirmation } = request.body;
+
+    const decoded = verify(token, authConfig.jwt.secret as string);
+    const { user } = decoded as TokenPayload;
+
+    const resetUserPassword = new ResetUserPasswordService();
+
+    const updated = await resetUserPassword.execute(
+      user.id,
+      newPassword,
+      newPasswordConfirmation,
+    );
+
+    // Return to login page
+    return response.redirect('/login');
```

Comment:
/login is a front-end route, unrelated to the server

Reply:
Maybe we can then redirect to a external IP set on environment variables?

Set on deployment to the URL of our frontend.

Or even yet
> How about we only return a response to the client, with a 204 No Content.

And treat redirects on the frontend.

Reply:
As in _don't know how to_ or _don't know if that's the way to go_?

Reply:
how to

Reply:
I am currently getting some weird `Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops` when trying to render the frontend, and tried most things.  
I have already coded a solution that worked on another branch (the error happens independently from it), and will push it anyway. Try it out for me. It will all go to town after React 18 anyway []s.